core/server/common/src/main/java/alluxio/underfs/UfsManager.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A class manages the UFS used by different worker services.
+ */
+@ThreadSafe
+public class UfsManager implements Closeable {
+
+  /**
+   * The key of the UFS cache.
+   */
+  private static class Key {
+    private final String mScheme;
+    private final String mAuthority;
+    private final Map<String, String> mProperties;
+
+    Key(AlluxioURI uri, Map<String, String> properties) {
+      mScheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase();
+      mAuthority = uri.getAuthority() == null ? "" : uri.getAuthority().toLowerCase();
+      mProperties = (properties == null || properties.isEmpty()) ? null : properties;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(mScheme, mAuthority, mProperties);
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object == this) {
+        return true;
+      }
+
+      if (!(object instanceof Key)) {
+        return false;
+      }
+
+      Key that = (Key) object;
+      return Objects.equal(mAuthority, that.mAuthority) && Objects
+          .equal(mProperties, that.mProperties) && Objects.equal(mScheme, that.mScheme);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+          .add("authority", mAuthority)
+          .add("scheme", mScheme)
+          .add("properties", mProperties).toString();
+    }
+  }
+
+  /** Maps from key to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+  /** Maps from mount id to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Long, UnderFileSystem> mMountIdToUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+
+  private UnderFileSystem mRootUfs = null;
+  protected final Closer mCloser;
+
+  /**
+   * Constructs the instance.
+   */
+  public UfsManager() {
+    mCloser = Closer.create();
+  }
+
+  /**
+   * Gets a UFS instance from the cache if exists. Otherwise, creates a new instance and adds
+   * that to the cache. Use this method only when you create new UFS instances.
+   *
+   * @param path the UFS path
+   * @param ufsConf the UFS configuration
+   * @return the UFS instance
+   */
+  public UnderFileSystem getOrCreate(String path, Map<String, String> ufsConf) {
+    Key key = new Key(new AlluxioURI(path), ufsConf);
+    UnderFileSystem cachedFs = mUnderFileSystemMap.get(key);
+    if (cachedFs != null) {
+      return cachedFs;
+    }
+    UnderFileSystem fs = UnderFileSystemRegistry.create(path, ufsConf);
+    UnderFileSystem racingFs = mUnderFileSystemMap.putIfAbsent(key, fs);
+    if (racingFs == null) {
+      mCloser.register(fs);
+      return fs;
+    }
+    try {
+      fs.close();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```

Comment:
yeah, i think i wrote this. I think it is better to swallow it and log the exception too